src/Servant/Utils/Links.hs

```diff
@@ -6,77 +7,135 @@
 {-# LANGUAGE FunctionalDependencies #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE UndecidableInstances #-}
--- | Type safe internal links.
+
+-- | Type safe generation of internal links.
+--
+-- Given an API with a few endpoints:
 --
--- Provides the function 'mkLink':
+-- >>> :set -XDataKinds -XTypeFamilies -XTypeOperators
+-- >>> import Servant.API
+-- >>> import Servant.Utils.Links
+-- >>> import Data.Proxy
+-- >>>
+-- >>>
+-- >>>
+-- >>> type Hello = "hello" :> Get Int
+-- >>> type Bye   = "bye"   :> QueryParam "name" String :> Delete
+-- >>> type API   = Hello :<|> Bye
+-- >>> let api = Proxy :: Proxy API
 --
--- @
---   type API = Proxy ("hello" :> Get Int
---                :<|> "bye" :> QueryParam "name" String :> Post Bool)
+-- It is possible to generate links that are guaranteed to be within 'API' with
+-- 'safeLink'. The first argument to 'safeLink' is a type representing the
+-- endpoint you would like to point to. This will need to end in a verb like
+-- Get, or Post. The second argument is the API in which you would like to
+-- ensure the endpoint is within. Further arguments may or may not be required
+-- depending on the type of the endpoint. If everything lines up you will get a
+-- 'URI' out the other end.
 --
---   api :: API
---   api = proxy
+-- You may omit 'QueryParam's and the like should you not want to provide them,
+-- but types which form part of the URL path like 'Capture' must be included.
+-- The reason you may want to omit 'QueryParam's is that safeLink is a bit
+-- magical: if parameters are included that could take input it will return a
+-- function that accepts that input and generates a link. This is best shown
+-- with an example. Here, a link is generated with no parameters:
 --
---   link1 :: Proxy ("hello" :> Get Int)
---   link1 = proxy
+-- >>> let hello = Proxy :: Proxy ("hello" :> Get Int)
+-- >>> print (safeLink hello api :: URI)
+-- hello
 --
---   link2 :: Proxy ("hello" :> Delete)
---   link2 = proxy
+-- If the API has an endpoint with parameters then we can generate links with
+-- or without those:
 --
---   mkLink link1 API  --  typechecks, returns 'Link "/hello"'
+-- >>> let with = Proxy :: Proxy ("bye" :> QueryParam "name" String :> Delete)
+-- >>> print $ safeLink with api "Hubert"
```

Comment:
Simply leave it out of the link you are trying to generate. If it is not needed as per IsElem then it will still typecheck.